pylint/pyreverse/writer.py

```diff
@@ -44,94 +57,104 @@ def write(self, diadefs):
             self.printer.emit_edge(
-                rel.from_object.fig_id, rel.to_object.fig_id, **self.pkg_edges
+                rel.from_object.fig_id,
+                rel.to_object.fig_id,
+                type_=EdgeType.USES,
             )
 
-    def write_classes(self, diagram):
+    def write_classes(self, diagram: ClassDiagram) -> None:
         """write a class diagram"""
         # sorted to get predictable (hence testable) results
-        for i, obj in enumerate(sorted(diagram.objects, key=lambda x: x.title)):
-            self.printer.emit_node(i, **self.get_values(obj))
-            obj.fig_id = i
+        for obj in sorted(diagram.objects, key=lambda x: x.title):
+            obj.fig_id = obj.node.qname()
+            type_ = NodeType.INTERFACE if obj.shape == "interface" else NodeType.CLASS
+            self.printer.emit_node(
+                obj.fig_id, type_=type_, properties=self.get_class_properties(obj)
+            )
         # inheritance links
         for rel in diagram.get_relationships("specialization"):
             self.printer.emit_edge(
-                rel.from_object.fig_id, rel.to_object.fig_id, **self.inh_edges
+                rel.from_object.fig_id,
+                rel.to_object.fig_id,
+                type_=EdgeType.INHERITS,
             )
         # implementation links
         for rel in diagram.get_relationships("implements"):
             self.printer.emit_edge(
-                rel.from_object.fig_id, rel.to_object.fig_id, **self.imp_edges
+                rel.from_object.fig_id,
+                rel.to_object.fig_id,
+                type_=EdgeType.IMPLEMENTS,
             )
         # generate associations
         for rel in diagram.get_relationships("association"):
             self.printer.emit_edge(
                 rel.from_object.fig_id,
                 rel.to_object.fig_id,
                 label=rel.name,
-                **self.association_edges,
+                type_=EdgeType.ASSOCIATION,
             )
 
-    def set_printer(self, file_name, basename):
+    def set_printer(self, file_name: str, basename: str) -> None:
         """set printer"""
         raise NotImplementedError
 
-    def get_title(self, obj):
+    def get_title(self, obj: DiagramEntity) -> str:
         """get project title"""
         raise NotImplementedError
 
-    def get_values(self, obj):
-        """get label and shape for classes."""
+    def get_package_properties(self, obj: PackageEntity) -> NodeProperties:
+        """get label and shape for packages."""
         raise NotImplementedError
 
-    def close_graph(self):
-        """finalize the graph"""
+    def get_class_properties(self, obj: ClassEntity) -> NodeProperties:
+        """get label and shape for classes."""
         raise NotImplementedError
 
+    def save(self) -> None:
+        """write to disk"""
+        self.printer.generate(self.file_name)
+
 
 class DotWriter(DiagramWriter):
```

Comment:
Maybe we could start with a first merge request that create the following architecture and move relevant part of the code without modifying anything:
```
pyreverse:
    writter:
        diagram_writter.py
        dot_writter.py
        vcg_writter.py
    printer:
        dot_printer.py
        vcg_printer.py
```
What do you think ?

Reply:
Thank you for reorganizing the code the diff is manageable now :)